File: native/extensions/anysdk/platform/android/src/org/cocos2dx/javascript/SDKWrapper.java
/****************************************************************************
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

package org.cocos2dx.javascript;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.opengl.GLSurfaceView;
import android.os.Bundle;
import android.util.Log;

public class SDKWrapper {
	private final static boolean PACKAGE_AS = true;
	private final static String TAG = "SDKWrapper";
	private final static String CLASS_PATH = "com.anysdk.framework.PluginWrapper";
	private static Class<?> mClass = null;
	private static Context mCtx = null;

	private static SDKWrapper mInstace = null;
	public static SDKWrapper getInstance() {
		if (null == mInstace){
			mInstace = new SDKWrapper();
			if (PACKAGE_AS) {
				try {
					mClass = Class.forName(CLASS_PATH);
				} catch (Exception e) {
					Log.e(TAG, "Can not find class: " + CLASS_PATH);
				}
			}
		}
		return mInstace;	
	}

	public void init(Context context) {
	    mCtx = context;
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("init", Context.class).invoke(mClass, context);
			} catch (Exception e) {
				e.printStackTrace();
			}
			SDKWrapper.nativeLoadAllPlugins();
		}
		
	}

	public Context getContext(){
	    return mCtx;
	}

	public void setGLSurfaceView(GLSurfaceView view) {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("setGLSurfaceView", GLSurfaceView.class).invoke(mClass, view);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
	
	public void onResume() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onResume").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onPause() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onPause").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onDestroy() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onDestroy").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onActivityResult", int.class, int.class, Intent.class).invoke(mClass, requestCode, resultCode, data);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onNewIntent(Intent intent) {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onNewIntent", Intent.class).invoke(mClass, intent);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onRestart() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onRestart").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}	
	}

	public void onStop() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onStop").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onBackPressed() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onBackPressed").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onConfigurationChanged(Configuration newConfig) {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onConfigurationChanged", Configuration.class).invoke(mClass, newConfig);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onRestoreInstanceState(Bundle savedInstanceState) {
			if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onRestoreInstanceState", Bundle.class).invoke(mClass, savedInstanceState);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onSaveInstanceState(Bundle outState) {
			if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onSaveInstanceState", Bundle.class).invoke(mClass, outState);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void onStart() {
		if (PACKAGE_AS) {
			try {
				if (null != mClass)
				    mClass.getMethod("onStart").invoke(mClass);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
	
	private static native void nativeLoadAllPlugins();
}

File: templates/ios/service/SDKWrapper.m
/****************************************************************************
 Copyright (c) 2017-2020 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/
#import "SDKWrapper.h"

@interface SDKWrapper ()

@property (nonatomic, strong) NSArray *serviceInstances;

@end

@implementation SDKWrapper

#pragma mark -
#pragma mark Singleton

static SDKWrapper *mInstace = nil;

+ (instancetype)shared {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        mInstace = [[super allocWithZone:NULL] init];
        [mInstace initSDKWrapper];
    });
    return mInstace;
}
+ (id)allocWithZone:(struct _NSZone *)zone {
    return [SDKWrapper shared];
}

+ (id)copyWithZone:(struct _NSZone *)zone {
    return [SDKWrapper shared];
}

#pragma mark -
#pragma mark private methods
- (void)initSDKWrapper {
    [self loadSDKClass];
}

- (void)loadSDKClass {
    NSMutableArray *sdks = [NSMutableArray array];
    @try {
        NSString *path = [NSString stringWithFormat:@"%@/service.json", [[NSBundle mainBundle] resourcePath]];
        NSData *data = [NSData dataWithContentsOfFile:path options:NSDataReadingMappedIfSafe error:nil];
        id obj = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        id dic = obj[@"serviceClasses"];
        if (dic == nil) @throw [[NSException alloc] initWithName:@"JSON Exception" reason:@"serviceClasses not found" userInfo:nil];
        for (NSString *str in dic) {
            NSString *className = [[str componentsSeparatedByString:@"."] lastObject];
            Class clazz = NSClassFromString(className);
            if (clazz == nil) @throw [[NSException alloc] initWithName:@"Cass Exception"
                                                            reason:[NSString stringWithFormat:@"class '%@' not found", className]
                                                          userInfo:nil];
            id sdk = [[clazz alloc] init];
            [sdks addObject:sdk];
        }
    } @catch (NSException *e) {
    }
    self.serviceInstances = [NSArray arrayWithArray:sdks];
}

#pragma mark -
#pragma mark Application lifecycle
- (void)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    for (id <SDKDelegate> sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(application:didFinishLaunchingWithOptions:)]) {
            [sdk application:application didFinishLaunchingWithOptions:launchOptions];
        }
    }
    
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    for (id sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(applicationDidBecomeActive:)]) {
            [sdk applicationDidBecomeActive:application];
        }
    }
}

- (void)applicationWillResignActive:(UIApplication *)application {
    for (id sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(applicationWillResignActive:)]) {
            [sdk applicationWillResignActive:application];
        }
    }
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
    for (id sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(applicationDidEnterBackground:)]) {
            [sdk applicationDidEnterBackground:application];
        }
    }
}

- (void)applicationWillEnterForeground:(UIApplication *)application {
    for (id sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(applicationWillEnterForeground:)]) {
            [sdk applicationWillEnterForeground:application];
        }
    }
}

- (void)applicationWillTerminate:(UIApplication *)application {
    for (id sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(applicationWillTerminate:)]) {
            [sdk applicationWillTerminate:application];
        }
    }
}

- (void)applicationDidReceiveMemoryWarning:(UIApplication *)application {
    for (id sdk in self.serviceInstances) {
        if ([sdk respondsToSelector:@selector(applicationDidReceiveMemoryWarning:)]) {
            [sdk applicationDidReceiveMemoryWarning:application];
        }
    }
}

@end
